specification/dartLangSpec.tex

```diff
@@ -3233,59 +3235,81 @@ \subsubsection{The Method \code{noSuchMethod}}
 \end{dartCode}
 
 
-\subsubsection{The Operator `=='}
+\subsubsection{The Operator `==' and Primitive Equality}
 \LMLabel{theOperatorEqualsEquals}
 
 \LMHash{}%
 The operator \lit{==} is used implicitly in certain situations,
 and in particular constant expressions
 (\ref{constants})
 give rise to constraints on that operator.
-In order to specify these constraints just once we introduce the notion of a
-% Neither \syntax nor \lit works, so we fall back to `\code{==}'.
-\IndexCustom{primitive operator `\code{==}'}{%
-  operator `\code{==}'!primitive}:
+The situation is similar with the getter \code{hashCode}.
+In order to specify these constraints just once we introduce the notion of
+\Index{primitive equality}.
+
+\rationale{%
+Certain constant expressions are known to have a value
+whose equality is primitive.
+This is useful to know because it allows
+the value of equality expressions
+and the value of invocations of \code{hashCode}
+to be computed at compile-time.
+In particular, this can be used to build
+constant collections at compile-time,
+and it can be used to check that
+all elements in a constant set are distinct,
+and all keys in a constant map are distinct.%
+}
 
 \begin{itemize}
-\item Every instance of type \code{int} and \code{String}
-  has a primitive operator \lit{==}.
+\item The null object has primitive equality
+  (\ref{null}).
+\item Every instance of type \code{bool}, \code{int}, and \code{String}
+  has primitive equality.
 \item Every instance of type \code{Symbol}
   which was originally obtained by evaluation of a literal symbol or
   a constant invocation of a constructor of the \code{Symbol} class
-  has a primitive operator \lit{==}.
-\item Every instance of type \code{Type}
+  has primitive equality.
+\item
+  %% TODO(eernst): With Dart 2.15, we need to mention `List<int>` here,
+  %% unless 'constant type literal' includes that case.
+  Every instance of type \code{Type}
   which was originally obtained by evaluating a constant type literal
   (\ref{dynamicTypeSystem})
-  has a primitive operator \lit{==}.
-\item An instance $o$ has a primitive operator \lit{==}
-  if the dynamic type of $o$ is a class $C$,
-  and $C$ has a primitive operator \lit{==}.
-\item The class \code{Object} has a primitive operator \lit{==}.
-\item A class $C$ has a primitive operator \lit{==}
-  if it does not have an implementation of the operator \lit{==}
-  that overrides the one inherited from \code{Object}.
-  \commentary{%
-  In particular, the following have a primitive operator \lit{==}:
-  The null object (\ref{null}),
-  function objects obtained by function closurization of
-  a static method or a top-level function
-  (\ref{functionClosurization}),
-  instances of type \code{bool}
-  (\ref{booleans}),
-  and instances obtained by evaluation of a list literal
+  has primitive equality.
+\item
+  Let $o$ be an object obtained by evaluation of a list literal
```

Comment:
You say "constant expressions" above, but not here.
I'd say "constant list literal", etc., just to avoid possible doubt.

Reply:
Done.